Mathlib/Data/List/Defs.lean

```diff
@@ -0,0 +1,637 @@
+  | [] => 0
+  | g :: [] => g
+  | g :: h :: t => g + -h + alternatingSum t
+#align list.alternating_sum List.alternatingSum
+
+/-- The alternating product of a list. -/
+def alternatingProd {G : Type _} [One G] [Mul G] [Inv G] : List G → G
+  | [] => 1
+  | g :: [] => g
+  | g :: h :: t => g * h⁻¹ * alternatingProd t
+#align list.alternating_prod List.alternatingProd
+
+#align list.partition_map List.partitionMap
+#align list.find List.find?
+
+/-- `mfind tac l` returns the first element of `l` on which `tac` succeeds, and
+fails otherwise. -/
+def findM {α} {m : Type u → Type v} [Monad m] [Alternative m] (tac : α → m PUnit) : List α → m α :=
+  List.firstM <| fun a => (tac a) $> a
+#align list.mfind List.findM
+
+/-- `mbfind' p l` returns the first element `a` of `l` for which `p a` returns
+true. `mbfind'` short-circuits, so `p` is not necessarily run on every `a` in
+`l`. This is a monadic version of `list.find`. -/
+def findM?'
+    {m : Type u → Type v}
+    [Monad m] {α : Type u}
+    (p : α → m (ULift Bool)) : List α → m (Option α)
+  | [] => pure none
+  | x :: xs => do
+    let ⟨px⟩ ← p x
+    if px then pure (some x) else findM?' p xs
+#align list.mbfind' List.findM?'
+
+#align list.mbfind List.findM?
+#align list.many List.anyM
+#align list.mall List.allM
+
+section
+
+variable {m : Type → Type v} [Monad m]
+
+/-- `orM xs` runs the actions in `xs`, returning true if any of them returns
+true. `orM` short-circuits, so if an action returns true, later actions are
+not run. -/
+def orM : List (m Bool) → m Bool :=
+  anyM id
+#align list.mbor List.orM
+
+/-- `andM xs` runs the actions in `xs`, returning true if all of them return
+true. `andM` short-circuits, so if an action returns false, later actions are
+not run. -/
+def andM : List (m Bool) → m Bool :=
+  allM id
+#align list.mband List.andM
+
+end
+
+#align list.foldr_with_index List.foldrIdx
+#align list.foldl_with_index List.foldlIdx
+#align list.find_indexes List.findIdxs
+#align list.indexes_values List.indexesValues
+#align list.indexes_of List.indexesOf
+
+section foldIdxM
+
+variable {m : Type v → Type w} [Monad m]
+
+/-- Monadic variant of `foldlIdx`. -/
+def foldlIdxM {α β} (f : ℕ → β → α → m β) (b : β) (as : List α) : m β :=
+  as.foldlIdx
+    (fun i ma b => do
+      let a ← ma
+      f i a b)
+    (pure b)
+#align list.mfoldl_with_index List.foldlIdxM
+
+/-- Monadic variant of `foldrIdx`. -/
+def foldrIdxM {α β} (f : ℕ → α → β → m β) (b : β) (as : List α) : m β :=
+  as.foldrIdx
+    (fun i a mb => do
+      let b ← mb
+      f i a b)
+    (pure b)
+#align list.mfoldr_with_index List.foldrIdxM
+
+end foldIdxM
+
+
+section mapIdxM
+
+-- porting notes: This was defined in `mathlib` with an `Applicative`
+-- constraint on `m` and have been `#align`ed to the `Std` versions defined
+-- with a with a `Monad` typeclass constraint
+-- since all `Monad`s are `Applicative` this won't cause issues
+-- downstream & `Monad`ic code is more performant per Mario C
```

Comment:
```suggestion
-- Since all `Monad`s are `Applicative` this won't cause issues
-- downstream & `Monad`ic code is more performant per Mario C.
```
Can you add the link to Mario's comment on zulip?